packages/transaction-controller/src/helpers/PendingTransactionTracker.ts

```diff
@@ -5,231 +5,471 @@ import EventEmitter from 'events';
+
+    const blocksSinceFirstRetry =
+      Number.parseInt(latestBlockNumber, 16) -
+      Number.parseInt(firstRetryBlockNumber, 16);
+
+    const retryCount = txMeta.retryCount || 0;
+
+    // Exponential backoff to limit retries at publishing
+    // Capped at ~15 minutes between retries
+    const requiredBlocksSinceFirstRetry = Math.min(
+      MAX_RETRY_BLOCK_DISTANCE,
+      Math.pow(2, retryCount),
+    );
+
+    return blocksSinceFirstRetry >= requiredBlocksSinceFirstRetry;
+  }
+
+  async #checkTransaction(txMeta: TransactionMeta) {
+    const { hash, id } = txMeta;
+
+    if (!hash) {
+      const error = new Error(
+        'We had an error while submitting this transaction, please try again.',
+      );
+
+      error.name = 'NoTxHashError';
+
+      this.#failTransaction(txMeta, error);
+
+      return;
+    }
+
+    if (this.#isNonceTaken(txMeta)) {
+      log('Nonce already taken', id);
+      this.#dropTransaction(txMeta);
+      return;
+    }
 
     try {
-      await Promise.all(
-        transactions.map(async (meta, index) => {
-          if (!meta.verifiedOnBlockchain && meta.chainId === currentChainId) {
-            const [reconciledTx, updateRequired] =
-              await this.#blockchainTransactionStateReconciler(meta);
-
-            if (updateRequired) {
-              transactions[index] = reconciledTx;
-              gotUpdates = updateRequired;
-            }
-          }
-        }),
+      const receipt = await this.#getTransactionReceipt(hash);
+      const isSuccess = receipt?.status === '0x1';
+
+      if (receipt && !isSuccess) {
+        log('Transaction receipt has failed status');
+
+        this.#failTransaction(
+          txMeta,
+          new Error('Transaction dropped or replaced'),
+        );
+
+        return;
+      }
+
+      if (receipt?.blockNumber && isSuccess) {
+        await this.#onTransactionConfirmed(txMeta, receipt);
+        return;
+      }
+    } catch (error: any) {
+      log('Failed to check transaction', id, error);
+
+      this.#warnTransaction(
+        txMeta,
+        error.message,
+        'There was a problem loading this transaction.',
       );
-    } catch (error) {
-      log('Error checking pending transactions', error);
+
+      return;
     }
 
-    /* istanbul ignore else */
-    if (gotUpdates) {
-      this.hub.emit('transactions', transactions);
+    if (await this.#isTransactionDropped(txMeta)) {
+      this.#dropTransaction(txMeta);
     }
   }
 
-  /**
-   * Method to verify the state of a transaction using the Blockchain as a source of truth.
-   *
-   * @param meta - The local transaction to verify on the blockchain.
-   * @returns A tuple containing the updated transaction, and whether or not an update was required.
-   */
-  async #blockchainTransactionStateReconciler(
-    meta: TransactionMeta,
-  ): Promise<[TransactionMeta, boolean]> {
+  async #onTransactionConfirmed(
+    txMeta: TransactionMeta,
+    receipt: TransactionReceipt,
```

Comment:
```suggestion
    receipt: TransactionReceipt & { blockNumber: string; blockHash: string },
```

Reply:
Rather than extend the type of the argument, I've created a `SuccessfulTransactionReceipt` type for clarity.

Reply:
Good call!